rfcs/20230324-graph-api-dynamic-shape/README.md

```diff
@@ -0,0 +1,376 @@
+shape scenario, users will not be able to compile a partition once ahead of time
+and use it repeatedly for the following iterations.
+
+## 2. Motivation
+
+The main purposes of this proposal are to:
+
+- Explore the possibility of compiling partitions with dynamic shape to reduce
+  the overhead of re-compilation.
+- Support the usage of "compile once, use in every iteration" from the
+  perspective of frameworks.
+- Extend the library operation set and programming model to support shape
+  tensors and shape manipulation operations which can enlarge the partition
+  subgraph and deliver more optimization opportunity to the library backends.
+- Expose the runtime dimension feature of primitive API through graph API.
+
+## 3. API proposals
+
+### 3.1 Graph building stage
+
+At the graph building stage, API users create a graph and add operations
+associated with input and output logical tensors to the graph. The users also
+need to specify the ID, data type, and dimension information for the input and
+output logical tensors.
+
+#### 3.1.1 Dimension representation
+
+With the existing Graph API, input and output shapes may be unknown at graph
+building stage but should be determined at partition compilation stage.
+Currently, INT64_MIN
+([DNNL_GRAPH_UNKNOWN_DIM](https://oneapi-src.github.io/oneDNN/group_dnnl_graph_api_logical_tensor.html#macros))
+is used to represent an unknown dimension, 0 is used to represent 0-size
+dimension, and positive values are used to represent explicit dimension sizes.
+For example, shape (INT64_MIN, INT64_MIN, 32, 32) means the shape has 4
+dimensions (`ndims = 4` in the definition of logical tensor) with the first two
+dimensions unknown. Shapes containing unknown dimensions can be used to define
+input and output logical tensors at graph building stage or the output tensors
+at partition compilation stage. For the latter case, the output shape should be
+deducible according to the shape inference rules of each operation.
+
+To extend the support to dynamic shape problem, this rfc proposes to define
+another wildcard value `DNNL_GRAPH_DYNAMIC_DIM` as the placeholder for dynamic
+dimensions. It brings two main benefits to differentiate "unknown dimensions"
+and "dynamic dimensions":
+
+1. It helps to keep the backward compatibility for INT64_MIN dimension as
+  unknown dimension.
+2. It allows to check dynamic dimension and skip the related ops graph
+   partitioning if a backend does not support dynamic dimension.
+
+| API                        | Value              | Description                                                   |
+|:---------------------------|:-------------------|:--------------------------------------------------------------|
+| DNNL_GRAPH_UNKNOWN_DIM     | INT64_MIN          | Unknown dimension but can be determined at compilation stage. |
+| **DNNL_GRAPH_DYNAMIC_DIM** | INT64_MIN + 1      | Dynamic dimension which is unknown until execution stage.     |
```

Comment:
Hi @eralmual, thanks for the review.

> This will allow users to create a tensor at graph creation time without having any information of it's shape or even number of dimensions, so at execution stage the specific shape for the tensors can be defined. Is my understanding correct?

**DNNL_GRAPH_DYNAMIC_DIM** is a placeholder to represent whether a dimension is dynamic or not. As mentioned in the rfc, number of dimensions (aka. ndims) is needed to allow check dynamic dimension at graph creation stage and skip partitioning if a backend cannot support dynamic shape compilation.

> This would really help oneDNN Graph integration into ONNX Runtime for the oneDNN EP, since most models don't provide **any** shape information of a tensor at graph creation time, so we don't get a shape or the number of dimensions, just the data type. It is only at execution time when the inputs have been provided and we know their shapes that we can propagate the information throughout the graph. This why we need to be able to specify the tensor shapes at execution in order to fully support dynamic shapes in ORT using oneDNN Graph.

I see your points. With the existing graph API, there are two backends as of now: primitive based (DNNL) and graph compiler. Different supporting capabilities of dynamic shape are provided in each backend. oneDNN Graph wants to filter out those ops with dynamic shape inputs early if cannot support. Please also take a look at below section `3.2 Graph partitioning stage`. 

BTW, in ONNXRT, how dynamic is the input shape in typical scenarios? Will some certain dimension or all dimensions be dynamic? Will ndims also be dynamic?

Reply:
> I see your points. With the existing graph API, there are two backends as of now: primitive based (DNNL) and graph compiler. Different supporting capabilities of dynamic shape are provided in each backend. oneDNN Graph wants to filter out those ops with dynamic shape inputs early if cannot support. Please also take a look at below section `3.2 Graph partitioning stage`.

Currently we are using the primitive based backend but the graph backed has some advantages like automated op fusion and has the chance to simplify quite a lot of our code so we are very interested on the Graph backend.

> BTW, in ONNXRT, how dynamic is the input shape in typical scenarios? Will some certain dimension or all dimensions be dynamic? Will ndims also be dynamic?

ORT has a few stages, the one where we _claim_ parts of the graph does not provide any kind of information about the input shape, essentially they just ask something like: Do you support Softmax with fp32 I/O? If we do, we claim the op, at this stage that's all we know.
Later on when we are going to execute our part of the graph, ORT provides the inputs to the operators we claimed, at this point because we have the input we can get it's shape.
This means that at graph creation time we don't even know the ndims of the tensors, I might also mention that while we don't know the number of ndims this does not mean it is dynamic, it's just unknown util execution time, this happens to models with dynamic dimensions which is pretty much every model we work with.